File: app/src/main/java/com/example/tetrisapp/data/local/dao/LeaderboardDao.java
package com.example.tetrisapp.data.local.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.example.tetrisapp.model.local.entity.LeaderboardEntry;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface LeaderboardDao {
    @Query("SELECT * FROM leaderboardentry WHERE hash = :hash")
    Single<LeaderboardEntry> getByHash(String hash);

    @Query("SELECT * FROM leaderboardentry WHERE hash IS NOT NULL")
    Single<List<LeaderboardEntry>> getWithHash();

    @Query("SELECT * FROM leaderboardentry WHERE hash IS NULL")
    Single<List<LeaderboardEntry>> getWithoutHash();

    @Query("SELECT * FROM leaderboardentry WHERE id=:id")
    Single<LeaderboardEntry> getById(long id);

    @Query("SELECT * FROM leaderboardentry")
    Maybe<List<LeaderboardEntry>> getAll();

    @Query("SELECT * FROM leaderboardentry WHERE uploaded = :uploaded AND hash IS NOT NULL")
    Maybe<List<LeaderboardEntry>> getByUploaded(boolean uploaded);

    @Query("SELECT * FROM leaderboardentry WHERE score = (SELECT MAX(score) FROM leaderboardentry) LIMIT 1")
    Single<LeaderboardEntry> getBest();

    @Query("SELECT * FROM leaderboardentry ORDER BY score DESC")
    Single<List<LeaderboardEntry>> getSorted();

    @Query("SELECT COUNT(*) FROM leaderboardentry")
    Single<Integer> getGamesCount();

    @Query("SELECT AVG(score) FROM leaderboardentry")
    Single<Integer> getAverageScore();

    @Query("SELECT AVG(level) FROM leaderboardentry")
    Single<Integer> getAverageLevel();

    @Query("SELECT AVG(lines) FROM leaderboardentry")
    Single<Integer> getAverageLines();

    @Query("SELECT MAX(score) FROM leaderboardentry")
    Single<Integer> getBestScore();

    @Query("SELECT MAX(level) FROM leaderboardentry")
    Single<Integer> getBestLevel();

    @Query("SELECT MAX(lines) FROM leaderboardentry")
    Single<Integer> getBestLines();

    @Insert
    Single<Long> insert(LeaderboardEntry leaderboardEntry);

    @Delete
    Single<Integer> delete(LeaderboardEntry entry);

    @Update
    Completable update(LeaderboardEntry entry);
}

File: app/src/main/java/com/example/tetrisapp/data/local/dao/LeaderboardDao.kt
package com.example.tetrisapp.data.local.dao

import androidx.room.*
import com.example.tetrisapp.model.local.LeaderboardEntry

@Dao
interface LeaderboardDao {
    @Query("SELECT * FROM leaderboardentry WHERE hash = :hash")
    fun getByHash(hash: String?): LeaderboardEntry?

    @Query("SELECT * FROM leaderboardentry WHERE hash IS NOT NULL")
    fun getWithHash(): List<LeaderboardEntry>?

    @Query("SELECT * FROM leaderboardentry WHERE hash IS NULL")
    fun getWithoutHash(): List<LeaderboardEntry>?

    @Query("SELECT * FROM leaderboardentry WHERE id=:id")
    fun getById(id: Long): LeaderboardEntry?

    @Query("SELECT * FROM leaderboardentry")
    fun getAll(): List<LeaderboardEntry>?

    @Query("SELECT * FROM leaderboardentry WHERE uploaded = :uploaded AND hash IS NOT NULL")
    fun getByUploaded(uploaded: Boolean): List<LeaderboardEntry>?

    @Query("SELECT * FROM leaderboardentry WHERE score = (SELECT MAX(score) FROM leaderboardentry) LIMIT 1")
    fun getBest(): LeaderboardEntry?

    @Query("SELECT * FROM leaderboardentry ORDER BY score DESC")
    fun getSorted(): List<LeaderboardEntry>?

    @Query("SELECT SUM(timeInGame) FROM leaderboardentry")
    fun getTimeInGame(): Int?

    @Query("SELECT COUNT(*) FROM leaderboardentry")
    fun getGamesCount(): Int?

    @Query("SELECT AVG(score) FROM leaderboardentry")
    fun getAverageScore(): Int?

    @Query("SELECT AVG(level) FROM leaderboardentry")
    fun getAverageLevel(): Int?

    @Query("SELECT AVG(lines) FROM leaderboardentry")
    fun getAverageLines(): Int?

    @Query("SELECT MAX(score) FROM leaderboardentry")
    fun getBestScore(): Int?

    @Query("SELECT MAX(level) FROM leaderboardentry")
    fun getBestLevel(): Int?

    @Query("SELECT MAX(lines) FROM leaderboardentry")
    fun getBestLines(): Int?

    @Insert
    fun insert(leaderboardEntry: LeaderboardEntry?): Long?

    @Delete
    fun delete(entry: LeaderboardEntry?)

    @Update
    fun update(entry: LeaderboardEntry?)
}